core/trino-main/src/main/java/io/trino/operator/TopNRankingOperator.java

```diff
@@ -291,35 +357,59 @@ public Page getOutput()
             unfinishedWork = null;
         }
 
-        if (!finishing) {
+        if (!finishing && (groupedTopNBuilder == null || !groupedTopNBuilder.isFull()) && outputIterator == null) {
             return null;
         }
 
-        if (outputIterator == null) {
+        if (outputIterator == null && groupedTopNBuilder != null) {
             // start flushing
             outputIterator = groupedTopNBuilder.buildResult();
         }
 
         Page output = null;
-        if (outputIterator.hasNext()) {
+        if (outputIterator != null && outputIterator.hasNext()) {
             // rewrite to expected column ordering
             output = outputIterator.next().getColumns(outputChannels);
         }
+        else {
+            closeGroupedTopNBuilder();
+            return null;
+        }
         updateMemoryReservation();
         return output;
     }
 
+    @Override
+    public void close()
+    {
+        closeGroupedTopNBuilder();
+    }
+
+    private void closeGroupedTopNBuilder()
+    {
+        outputIterator = null;
+        groupedTopNBuilder = null;
+        localMemoryContext.setBytes(0);
+    }
+
     @VisibleForTesting
     public int getCapacity()
     {
-        checkState(groupByHash != null);
-        return groupByHash.getCapacity();
+        return groupedTopNBuilder == null ? 0 : groupedTopNBuilder.getCapacity();
     }
 
     private boolean updateMemoryReservation()
     {
+        if (groupedTopNBuilder == null) {
+            localMemoryContext.setBytes(0);
+            return true;
+        }
         // TODO: may need to use trySetMemoryReservation with a compaction to free memory (but that may cause GC pressure)
-        localUserMemoryContext.setBytes(groupedTopNBuilder.getEstimatedSizeInBytes());
+        localMemoryContext.setBytes(groupedTopNBuilder.getEstimatedSizeInBytes());
+        if (groupedTopNBuilder.isFlushable()) {
+            // do not yield on memory for partial aggregations
+            return true;
+        }
```

Comment:
@sopel39, I noticed that we do this in the HashAggregationOperator as well. Why do we not want to yield on memory for partial aggregations? It seems more semantically correct to block on memory, even for partials.

Reply:
If the partial is about to flush, you _must_ let the memory request succeed instead of blocking because otherwise you won't be able to transition into the flushing state to release memory which creates an easy opportunity for queries to deadlock on each other releasing memory and not making forward progress.

Reply:
@pettyjamesm, what about the cases where the partial is not about to flush? Should it still continue to consume more and more memory?

Reply:
> @pettyjamesm, what about the cases where the partial is not about to flush? Should it still continue to consume more and more memory?

This is based on https://github.com/trinodb/trino/commit/561deca0c7e6877d13873fd8696941477c4fe036. TBH I don't recall reason for it. Maybe we could yield for partial aggregation too. Someone would have to run tests with this enabled.

Reply:
Should it be a part of this PR? Since HashAggregationOperator does the same thing, it would be better to be handled separately.

Reply:
> @pettyjamesm, what about the cases where the partial is not about to flush? Should it still continue to consume more and more memory?

@erichwang - that case is a little trickier to reason about. I think the current strategy employed by `HashAggregationOperator` which this PR is borrowing the approach from behaves the way it does because of a series of compromises / workarounds for race conditions and deadlocks over time. The idea seems to be that using system memory and effectively guaranteeing all operator instances in leaf split pipelines access to the memory limit (with more allowances for the transition to flushing) is better than the alternative of blocking. This might also be related to the fact that memory pressure will not backpressure new splits from starting, so pausing leaf split pipelines on memory will only accelerator the node OOM condition. Hard to say, but there would be risks associated with going the other direction on this operator without some other significant modifications and testing.